Add UI for modifying chunk options

kevinushey opened the pull request:
Not quite ready for merge; needs discussion to refine UI + available options.

This PR adds a 'gear' icon that allows users to modify chunk options more easily:

![chunk-options](https://cloud.githubusercontent.com/assets/1976582/7755673/d751862e-ffaa-11e4-9745-745e3e20bc34.gif)

The idea is that changes in the popup immediately synchronize with the displayed text.

## TODO
- [x] Fix focus issues when clicking a gear when source editor not in focus
- [x] Consider other, non-boolean options (figure width / height? engine?)
- [x] Consider three-state checkbox to allow for [empty | off | on] states
- [x] Styling?
- [x] Display over text

kevinushey commented:
Here's how it looks now, with placeholder styling for the tri-state checkbox:

![screen shot 2015-05-21 at 3 31 08 pm](https://cloud.githubusercontent.com/assets/1976582/7760257/728b7c9e-ffce-11e4-9095-750ba7b6d863.png)

I also added a tiny bit of styling to the icons (just so they have their own background where they're displayed)

After I fix up the focus issue mentioned earlier + get some new images / styling for the icons I think it should be good for merge.

jjallaire commented:
This is looking very good!

I'd change the text to (removing all of the question marks as they are
implied):

Evaluate R code

Print R output into document

Include R code in document



On Thu, May 21, 2015 at 6:32 PM, Kevin Ushey <notifications@github.com>
wrote:

> Here's how it looks now, with placeholder styling for the tri-state
> checkbox:
>
> [image: screen shot 2015-05-21 at 3 31 08 pm]
> <https://cloud.githubusercontent.com/assets/1976582/7760257/728b7c9e-ffce-11e4-9095-750ba7b6d863.png>
>
> I also added a tiny bit of styling to the icons (just so they have their
> own background where they're displayed)
>
> After I fix up the focus issue mentioned earlier + get some new images /
> styling for the icons I think it should be good for merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/428#issuecomment-104440704>.
>

kevinushey commented:
Almost ready to go; just need to fix one more bug re: display here:

![screen shot 2015-05-21 at 4 13 58 pm](https://cloud.githubusercontent.com/assets/1976582/7760865/7fe9e5a0-ffd4-11e4-8c0f-882dcf30ee8b.png)

I think I'll just make the background opaque for the icons and pick an appropriate color depending on the theme.

jjallaire commented:
Just sent Paul a note regarding our outstanding UI questions so hopefully we'll have that resolved and incorporated today. Will merge after we get those tweaks in. Note that I still recommend that we tweak the language. Here's a variation that uses the actual chunk option as the first word of each sentence:

```
Evaluate R code
Include chunk output in document
Echo R code into document
```

jjallaire commented on src/gwt/src/org/rstudio/studio/client/RStudioGinModule.java:
```diff
@@ -365,6 +366,8 @@ protected void configure()
 
       bind(DocDisplay.class).to(AceEditor.class);
       
+      bind(ChunkIconsManager.class).in(Singleton.class);
```
Not sure that this needs to be a Singleton (is there state we need to share across invocations?).

jjallaire commented on src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkIconsManager.java:
```diff
@@ -0,0 +1,211 @@
+{
+   public ChunkIconsManager()
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
```
Perhaps it's this event handler that's the "shared" element which pushes this into being a singleton?

kevinushey replied:
Yeah, that's basically why -- not sure if this is the best way of handling it (e.g. in the future when we have multiple source documents / windows open, should we have one instance per window?) but I think it's reasonable, at least for now.

gokceneraslan commented:
How about displaying the gear icon of a chunk only if the mouse is hovering over the area of the chunk in question?

kevinushey commented:
I think we want it visible by default for now (just to increase exposure to users) but we should definitely have an option for hiding / intelligently displaying the toolbar.

kevinushey commented:
Current treatment:

![screen shot 2015-05-22 at 12 34 26 pm](https://cloud.githubusercontent.com/assets/1976582/7778316/f0625448-007e-11e5-9fe8-31f948835389.png)

@jjallaire, I didn't use the `ThemedButton`s because they were to large to fit comfortably here, and it doesn't look like there's a simple way to resize.

There's also a subtle issue where the background colouring of our themed popup panel doesn't work great -- this is because the border includes an opaque white, and that white overlays the background colour we attempt to set.

Do you think it's worth attempting to fix that? (setting white to transparent; explicitly declaring background colors elsewhere as needed). That could definitely get messy though.

jjallaire commented:
Are you using SmallButton? I think that will give you the size /
form-factor you want and still be themed consistently across
browsers/platforms.

We should probably talk realtime about the issues surrounding the
background coloring. This might be worth addressing however not if as you
said gets too messy.

I think we should change the text of the options to read as follows (note
the first word of each sentence matches the chunk option):

Evaluate R code
Include chunk output
Echo R code

If you want them to be a bit longer you can use:

Evaluate R code
Include chunk output in document
Echo R code in document

But if the first set looks okay I'd go with that.




On Fri, May 22, 2015 at 3:36 PM, Kevin Ushey <notifications@github.com>
wrote:

> Current treatment:
>
> [image: screen shot 2015-05-22 at 12 34 26 pm]
> <https://cloud.githubusercontent.com/assets/1976582/7778316/f0625448-007e-11e5-9fe8-31f948835389.png>
>
> @jjallaire <https://github.com/jjallaire>, I didn't use the ThemedButtons
> because they were to large to fit comfortably here, and it doesn't look
> like there's a simple way to resize.
>
> There's also a subtle issue where the background colouring of our themed
> popup panel doesn't work great -- this is because the border includes an
> opaque white, and that white overlays the background colour we attempt to
> set.
>
> Do you think it's worth attempting to fix that? (setting white to
> transparent; explicitly declaring background colors elsewhere as needed).
> That could definitely get messy though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/428#issuecomment-104749926>.
>

kevinushey commented:
Okay, I think this is good enough for merge (there may be some behaviour refinement still needed but I think it's ready for master + initial testing in the dailies)

I added a simple `MiniPopupPanel` class that acts like the `MiniDialogPopupPanel`, but without the dialog header + the minimal styling we discussed. I also added some 'smart' drag handling so you can click popups any place without an 'input' handler and drag -- @jjallaire, do you want to sanity-check my implementation?

Here's what it looks like now:

![screen shot 2015-05-22 at 1 50 56 pm](https://cloud.githubusercontent.com/assets/1976582/7780888/e9662ff0-0099-11e5-90c6-8cc44f6d5549.png)

I also added a display option to toggle display of chunk icons, and added some color utilities for managing the background color of the inline toolbar.

kevinushey commented:
Redirect for the Chunk Options help link: https://github.com/rstudio/rstudio-web/pull/4
